lib/vnet/app_resolver.go

```diff
@@ -0,0 +1,177 @@
+// handled.
+//
+// [appProvider] is also used to get app certificates used to dial the apps.
+func NewTCPAppResolver(appProvider AppProvider) *TCPAppResolver {
+	return &TCPAppResolver{
+		appProvider: appProvider,
+		slog:        slog.With(teleport.ComponentKey, "VNet.AppResolver"),
+	}
+}
+
+// ResolveTCPHandler resolves a fully-qualified domain name (FQDN) to a TCPHandler for handling TCP
+// connections to a Teleport TCP app. It returns the TCPHandler if a matching app is found, else it will
+// return with match == false. Errors must only be returned for truly unexpected errors that should cause a
+// DNS request to fail.
+func (r *TCPAppResolver) ResolveTCPHandler(ctx context.Context, fqdn string) (handler TCPHandler, match bool, err error) {
+	profileNames, err := r.appProvider.ListProfiles()
+	if err != nil {
+		return nil, false, trace.Wrap(err, "listing profiles")
+	}
+	for _, profileName := range profileNames {
+		if fqdn == fullyQualify(profileName) {
+			// This is a query for the proxy address, which we'll never want to handle.
+			return nil, false, nil
+		}
+		if !isSubdomain(fqdn, profileName) {
+			// TODO(nklaassen): support leaf clusters and custom DNS zones.
+			continue
+		}
+
+		slog := r.slog.With("profile", profileName, "fqdn", fqdn)
+		rootClient, err := r.appProvider.GetCachedClient(ctx, profileName, "")
+		if err != nil {
+			// The user might be logged out from this one cluster (and retryWithRelogin isn't working). Don't
+			// return an error so that DNS resolution will be forwarded upstream instead of failing, to avoid
+			// breaking e.g. web app access (we don't know if this is a web or TCP app yet because we can't
+			// log in).
+			slog.InfoContext(ctx, "Failed to get teleport client.", "error", err)
+			continue
+		}
+		return r.resolveTCPHandlerForCluster(ctx, slog, rootClient.CurrentCluster(), profileName, "", fqdn)
+	}
+
+	// fqdn did not match any profile, forward the request upstream.
+	return nil, false, nil
+}
+
+func (r *TCPAppResolver) resolveTCPHandlerForCluster(
+	ctx context.Context,
+	slog *slog.Logger,
+	clt apiclient.GetResourcesClient,
+	profileName, leafClusterName, fqdn string,
+) (handler TCPHandler, match bool, err error) {
+	// An app public_addr could technically be full-qualified or not, match either way.
+	expr := fmt.Sprintf(`(resource.spec.public_addr == "%s" || resource.spec.public_addr == "%s") && hasPrefix(resource.spec.uri, "tcp://")`,
+		strings.TrimSuffix(fqdn, "."), fqdn)
+	resp, err := apiclient.GetResourcePage[types.AppServer](ctx, clt, &proto.ListResourcesRequest{
+		ResourceType:        types.KindAppServer,
+		PredicateExpression: expr,
+		Limit:               1,
+	})
+	if err != nil {
+		// Don't return an error so we can try to find the app in different clusters or forward the request upstream.
+		slog.InfoContext(ctx, "Failed to list application servers.", "error", err)
```

Comment:
This fails to prompt for relogin in certain conditions.

It works fine if VNet is used like this:

1. The user logs in and launches VNet.
2. They open the app through VNet, thus VNet now knows about the app.
3. The cert expires.
4. The user tries to open the app again. VNet forwards the request to the local proxy and the local proxy notices that the cert has expired.

However, if the user tries to connect to the app only after the cert has expired, then VNet does not know about the app and it enters this branch here. VNet forwards the request to the regular nameserver and the connection circumvents VNet completely, resulting in an unexpected behavior.

Reply:
The bad thing about this is that I believe the response to the forwarded request is cached, so even after the user refreshes the cert, they won't be able to connect to the app.

Off the top of my head, I don't know if this caching is on our side or the system's.

Reply:
It seems that it's cached on our side, because after restarting VNet everything works again.